code/modules/chemistry/Reagents-Medical.dm

```diff
@@ -102,23 +100,42 @@ datum
 				if(!M) M = holder.my_atom
 				if(!counter) counter = 1
 				//don't do shit if there's naloxone in you
-				if(!holder.has_reagent("naloxone"))
-					M.jitteriness = max(M.jitteriness-25,0)
-					if(holder.has_reagent("omegazine"))
-						holder.remove_reagent("omegazine", 3 * mult)
-
-					switch(counter += 1 * mult)
-						if(1 to 15)
-							if(probmult(7)) M.emote("yawn")
-						if(16 to 35)
-							M.drowsyness  = max(M.drowsyness, 20)
-						if(36 to INFINITY)
-							M.setStatus("paralysis", max(M.getStatusDuration("paralysis"), 3 SECONDS * mult))
-							M.drowsyness  = max(M.drowsyness, 20)
+				if(holder.has_reagent("naloxone"))
+					..()
+					return
+
+				if (M.reagents.get_reagent_amount("morphine") > max_volume)
+					max_volume = M.reagents.get_reagent_amount("morphine")
+					if (max_volume > overdose) // Separates the KO-y part of morphine from the heal-y
+						if (counter < 35) counter = 35
+
+				M.jitteriness = max(M.jitteriness-25,0)
+				if(holder.has_reagent("omegazine"))
+					holder.remove_reagent("omegazine", 3 * mult)
+
+				switch(counter += 1 * mult)
+					if(1 to 40)
+						if(probmult(5)) M.emote("yawn")
+					if(40 to 60)
+						if(probmult(25)) M.drowsyness  = max(M.drowsyness, 5)
+					if(60 to INFINITY)
+						M.setStatus("paralysis", max(M.getStatusDuration("paralysis"), 3 SECONDS * mult))
+						M.drowsyness  = max(M.drowsyness, 20)
+						holder.remove_reagent("morphine", 0.2 * mult) //morphine gets flushed faster while you are KO'd
+
+						if(probmult(35) && counter > 70)
+							M.losebreath += (5)
+							M.take_toxin_damage(2)
+
+				if (ishuman(M))
+					if (counter > 10 && counter < 60 && probmult(10)) // Has a chance to stop shock
+						var/mob/living/carbon/human/H = M
+						H.cure_disease_by_path(/datum/ailment/malady/shock)
```

Comment:
Diseases (or at least maladies) can handle reagent cures like this already, just add morphine to shock's reagentcure list